packages/nimble-components/src/table-column/mixins/custom-sort-order.ts

```diff
@@ -0,0 +1,96 @@
+// prettier-ignore
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type CustomSortOrderTableColumnConstructor<TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>> = abstract new (...args: any[]) => CustomSortOrderTableColumn<TColumnValidator>;
+
+// As the returned class is internal to the function, we can't write a signature that uses is directly, so rely on inference
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
+export function mixinCustomSortOrderColumnAPI<
+    TBase extends CustomSortOrderTableColumnConstructor<TColumnValidator>,
+    TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>
+>(base: TBase) {
+    /**
+     * The Mixin that provides a concrete column with the API to allow sorting
+     * by a field other than the default for that column.
+     */
+    abstract class CustomSortOrderColumn extends base {
+        public sortByFieldName?: string;
+
+        /** @internal */
+        public notifier?: Notifier;
+
+        /** @internal */
+        public sortByFieldNameChanged(): void {
+            this.updateOperandDataRecordFieldName();
+
+            if (typeof this.sortByFieldName === 'string' && !this.notifier) {
+                this.notifier = Observable.getNotifier(this.columnInternals);
+                this.notifier.subscribe(this);
+                this.updateCustomColumnSortingValidity();
+            } else {
+                this.notifier?.unsubscribe(this);
+                this.notifier = undefined;
+                this.updateCustomColumnSortingValidity();
+            }
+        }
+
+        /** @internal */
+        public updateOperandDataRecordFieldName(): void {
+            if (typeof this.sortByFieldName === 'string') {
+                this.columnInternals.operandDataRecordFieldName = this.sortByFieldName;
+                this.columnInternals.sortOperation = TableColumnSortOperation.basic;
+            } else {
+                this.columnInternals.operandDataRecordFieldName = this.getDefaultSortFieldName();
+                this.columnInternals.sortOperation = this.getDefaultSortOperation();
+            }
+        }
+
+        /** @internal */
+        public handleChange(_source: unknown, args: unknown): void {
+            if (args === 'groupingDisabled' || args === 'groupIndex') {
+                this.updateCustomColumnSortingValidity();
+            } else if (args === 'dataRecordFieldNames') {
+                this.updateOperandDataRecordFieldName();
+            }
+        }
+
+        /** @internal */
+        public updateCustomColumnSortingValidity(): void {
+            const hasCustomColumnSorting = typeof this.sortByFieldName === 'string';
+            const isGrouped = !this.columnInternals.groupingDisabled
+                && typeof this.columnInternals.groupIndex === 'number';
```

Comment:
Do we really need the groupIndex check? Can't we just check if groupingDisabled is false?

Reply:
Sure, I can make that change. That's probably more robust since it will catch any validation errors if the column is groupable, not just if it happened to be grouped at the time.